Design.md

```diff
@@ -0,0 +1,107 @@
+Fragment name | Resulting fragment key stored on hyperdrive must contain the parent, S3 object name. Or at least a prefix if too long. Resulting key must also be unique hyperdrive-wise.
+Fragment location | Each fragment must enable the hdclient to select the correct hyperdrive
+Fragment type |Each fragment should be tagged as data or coding (reading would then first prefer data parts since RS is systematic)
+
+Since we can specify which key to use on an hyperdrive, we could define a new key scheme to encode fragments in a unified key scheme. The idea is to minimize the amount of metadata we have to store with every object in the MongoDB database that can be used to locate the fragments. Note that we plan to use a 2+1 or 4+2 erasure code for now so the overhead of storing the full uuids for fragment location is not yet important. All fragment IDs should start with the same key as the object (part) key. If the object (part) key is indexed we can quickly find to which object belong the missing fragment (e.g. for repair, see below). The hdclient backend in cloud server will return this constructed key to cloud server main code.
+
+Proposed key generation scheme:
+* <genkey> := <version>#<topology_version>#<split>#<rep_policy>#<object_key>#<rand>#<location>[#<location>]
+* <version> := Natural (so 0 or 1 to start)
+* <topology_version> := Natural (so 0 or 1 to start) - if we use indirection instead of storing raw UUIDS we must versioned the mapping
+* <split> := <n_chunk>,<split_size>
+* <n_chunk> := number of splitted parts (1 for non-splitted objects)
+* <split_size> := size of each splitted parts, except last one (see hyperdrive keys below)
+* <rep_policy> := RS,k,m or CP,n (for n-1 copies)
+* <object_key> := parent S3 object key (or a prefix) - can contain anything but ‘#’
+* <rand> := 64 bits random number (unicity inside 1 hyperdrive)
+* <location>:= hyperdrive location (UUID, idx in table?, ip:port, ...)
+
+The fragments do not store the exact, really used hyperdrive keys. Those can be derived easily with only the generated key, even for split.
+* <stored_fragment_key> := <object_key>-<rand>-<start_offset>-<type>-<fragid>
+* <object_key> and <rand> are the ones defined above
+* <type> := ‘d’ for data, ‘c’ for coding
+* <fragid>:= index in main key fragment list
+* <start_offset> := used for splits. All split chunks share the same prefix, storing the offset is used to easily have range queries and avoid storing them all in the main key.
+
+Meets all requirements (except perhaps around the last stripe size of an ECN chunk). Split is required to contact the same hyperdrives for the object. This has several benefits: only selecting once, object located on only a handful of hyperdrives => less machines to contact = less opportunities to fail or hit a straggler. The way split is handled also enables us not to have a manifest, worries about its size, location or freshness.
```

Comment:
So a split object is always stored on only one hyperdrive, then ?
Or can it be split between multiple hyperdrives ? (In which case I have a hard time understanding how this number is determined, given the description of the optimization)